File: frontend/src/services/workspaces.ts
import { apiFetch } from '@/services/api';

export interface Workspace {
  id: string;
  name: string;
  slug: string;
  description: string;
  plan: 'free' | 'pro' | 'enterprise';
  is_active: boolean;
  max_experiments: number;
  max_feature_flags: number;
  max_members: number;
  member_count: number;
  experiment_count?: number;
  feature_flag_count?: number;
  api_key_count?: number;
  created_at: string;
  owner_username?: string;
}

export interface WorkspaceMember {
  user_id: string;
  username: string;
  email: string;
  role: 'OWNER' | 'ADMIN' | 'DEVELOPER' | 'ANALYST' | 'VIEWER';
  joined_at: string;
}

export interface WorkspaceAPIKey {
  id: string;
  name: string;
  key_prefix: string;
  scopes: string[];
  is_active: boolean;
  last_used_at: string | null;
  expires_at: string | null;
  created_at: string;
}

export interface CreateWorkspaceAPIKeyResponse extends WorkspaceAPIKey {
  key: string;
}

export interface WorkspaceInvite {
  token: string;
  workspace_name: string;
  inviter_username: string;
  email: string;
  role: string;
  expires_at: string;
}

export const workspaceService = {
  list: () =>
    apiFetch<Workspace[]>('/api/v1/workspaces/'),

  create: (data: Partial<Workspace>) =>
    apiFetch<Workspace>('/api/v1/workspaces/', { method: 'POST', json: data }),

  get: (id: string) =>
    apiFetch<Workspace>(`/api/v1/workspaces/${id}`),

  update: (id: string, data: Partial<Workspace>) =>
    apiFetch<Workspace>(`/api/v1/workspaces/${id}`, { method: 'PUT', json: data }),

  delete: (id: string) =>
    apiFetch<void>(`/api/v1/workspaces/${id}`, { method: 'DELETE' }),

  listMembers: (id: string) =>
    apiFetch<WorkspaceMember[]>(`/api/v1/workspaces/${id}/members`),

  addMember: (id: string, data: { user_id: string; role: string }) =>
    apiFetch<WorkspaceMember>(`/api/v1/workspaces/${id}/members`, { method: 'POST', json: data }),

  updateMember: (id: string, userId: string, role: string) =>
    apiFetch<WorkspaceMember>(`/api/v1/workspaces/${id}/members/${userId}`, {
      method: 'PUT',
      json: { role },
    }),

  removeMember: (id: string, userId: string) =>
    apiFetch<void>(`/api/v1/workspaces/${id}/members/${userId}`, { method: 'DELETE' }),

  sendInvite: (id: string, email: string, role: string) =>
    apiFetch<WorkspaceInvite>(`/api/v1/workspaces/${id}/invites`, {
      method: 'POST',
      json: { email, role },
    }),

  getInvite: (token: string) =>
    apiFetch<WorkspaceInvite>(`/api/v1/workspaces/invites/${token}`),

  acceptInvite: (token: string) =>
    apiFetch<{ workspace_id: string }>(`/api/v1/workspaces/invites/${token}/accept`, {
      method: 'POST',
    }),

  listAPIKeys: (id: string) =>
    apiFetch<WorkspaceAPIKey[]>(`/api/v1/workspaces/${id}/api-keys`),

  createAPIKey: (id: string, data: { name: string; scopes: string[] }) =>
    apiFetch<CreateWorkspaceAPIKeyResponse>(`/api/v1/workspaces/${id}/api-keys`, {
      method: 'POST',
      json: data,
    }),

  revokeAPIKey: (id: string, keyId: string) =>
    apiFetch<void>(`/api/v1/workspaces/${id}/api-keys/${keyId}`, { method: 'DELETE' }),

  rotateAPIKey: (id: string, keyId: string) =>
    apiFetch<CreateWorkspaceAPIKeyResponse>(
      `/api/v1/workspaces/${id}/api-keys/${keyId}/rotate`,
      { method: 'POST' }
    ),
};

File: modules/backend/app/schemas/workspaces.py
"""
Pydantic schemas for EP-057: Multi-Tenant Team Workspaces.
"""

import re
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, field_validator

# ─────────────────────────────────────────────────────────────────────────────
# Workspace
# ─────────────────────────────────────────────────────────────────────────────

_SLUG_RE = re.compile(r"^[a-z0-9][a-z0-9\-]{1,48}[a-z0-9]$")


class CreateWorkspaceRequest(BaseModel):
    """Payload for creating a new workspace."""

    name: str
    slug: str
    description: str = ""
    plan: str = "free"

    @field_validator("slug")
    @classmethod
    def validate_slug(cls, v: str) -> str:
        if not _SLUG_RE.match(v):
            raise ValueError(
                "Slug must be 3–50 characters, lowercase alphanumeric with hyphens, "
                "and must not start or end with a hyphen."
            )
        return v

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("name must not be empty")
        if len(v) > 200:
            raise ValueError("name must be at most 200 characters")
        return v

    @field_validator("plan")
    @classmethod
    def validate_plan(cls, v: str) -> str:
        allowed = {"free", "pro", "enterprise"}
        if v not in allowed:
            raise ValueError(f"plan must be one of {sorted(allowed)}")
        return v


class UpdateWorkspaceRequest(BaseModel):
    """Payload for updating an existing workspace (all fields optional)."""

    name: Optional[str] = None
    description: Optional[str] = None
    plan: Optional[str] = None


class WorkspaceResponse(BaseModel):
    """Response schema for a workspace."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    slug: str
    description: Optional[str]
    plan: str
    is_active: bool
    max_experiments: int
    max_feature_flags: int
    max_members: int
    max_api_keys: int
    created_at: datetime
    updated_at: datetime

    @field_validator("id", mode="before")
    @classmethod
    def coerce_id(cls, v):
        return str(v)

    @field_validator("plan", mode="before")
    @classmethod
    def coerce_plan(cls, v):
        if hasattr(v, "value"):
            return v.value
        return str(v)


class WorkspaceWithStatsResponse(WorkspaceResponse):
    """Workspace response enriched with live resource counts."""

    member_count: int = 0
    experiment_count: int = 0
    flag_count: int = 0
    api_key_count: int = 0


# ─────────────────────────────────────────────────────────────────────────────
# Members
# ─────────────────────────────────────────────────────────────────────────────


class AddMemberRequest(BaseModel):
    """Payload for adding an existing user to a workspace."""

    user_id: str
    role: str = "VIEWER"

    @field_validator("role")
    @classmethod
    def validate_role(cls, v: str) -> str:
        allowed = {"OWNER", "ADMIN", "DEVELOPER", "ANALYST", "VIEWER"}
        if v.upper() not in allowed:
            raise ValueError(f"role must be one of {sorted(allowed)}")
        return v.upper()


class UpdateMemberRoleRequest(BaseModel):
    """Payload for changing a member's role."""

    role: str

    @field_validator("role")
    @classmethod
    def validate_role(cls, v: str) -> str:
        allowed = {"OWNER", "ADMIN", "DEVELOPER", "ANALYST", "VIEWER"}
        if v.upper() not in allowed:
            raise ValueError(f"role must be one of {sorted(allowed)}")
        return v.upper()


class WorkspaceMemberResponse(BaseModel):
    """Response schema for a workspace member."""

    model_config = ConfigDict(from_attributes=True)

    workspace_id: str
    user_id: str
    role: str
    joined_at: datetime

    # Denormalised user fields (populated manually in endpoints)
    username: Optional[str] = None
    email: Optional[str] = None

    @field_validator("workspace_id", "user_id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)

    @field_validator("role", mode="before")
    @classmethod
    def coerce_role(cls, v):
        if hasattr(v, "value"):
            return v.value
        return str(v)


# ─────────────────────────────────────────────────────────────────────────────
# Invites
# ─────────────────────────────────────────────────────────────────────────────


class CreateInviteRequest(BaseModel):
    """Payload for sending a workspace invite."""

    email: str
    role: str = "VIEWER"

    @field_validator("role")
    @classmethod
    def validate_role(cls, v: str) -> str:
        # OWNER cannot be granted via invite
        allowed = {"ADMIN", "DEVELOPER", "ANALYST", "VIEWER"}
        if v.upper() not in allowed:
            raise ValueError(
                f"role must be one of {sorted(allowed)} (OWNER cannot be invited)"
            )
        return v.upper()


class WorkspaceInviteResponse(BaseModel):
    """Response schema for a workspace invite."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    workspace_id: str
    email: str
    role: str
    token: str
    expires_at: datetime
    accepted_at: Optional[datetime]

    @field_validator("id", "workspace_id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)

    @field_validator("role", mode="before")
    @classmethod
    def coerce_role(cls, v):
        if hasattr(v, "value"):
            return v.value
        return str(v)


# ─────────────────────────────────────────────────────────────────────────────
# API Keys
# ─────────────────────────────────────────────────────────────────────────────


class CreateAPIKeyRequest(BaseModel):
    """Payload for creating a new workspace-scoped API key."""

    name: str
    scopes: List[str] = ["flags:read", "experiments:read", "track:write"]
    expires_at: Optional[datetime] = None


class CreateAPIKeyResponse(BaseModel):
    """Response schema returned exactly once when a new API key is created."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    key_prefix: str
    key: str  # plaintext — shown ONCE only
    scopes: List[str]
    created_at: datetime

    @field_validator("id", mode="before")
    @classmethod
    def coerce_id(cls, v):
        return str(v)


class WorkspaceAPIKeyResponse(BaseModel):
    """Safe response schema for listing API keys (no plaintext)."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    key_prefix: str
    scopes: List[str]
    is_active: bool
    last_used_at: Optional[datetime]
    expires_at: Optional[datetime]
    created_at: datetime

    @field_validator("id", mode="before")
    @classmethod
    def coerce_id(cls, v):
        return str(v)
